### ISSUE TITLE
[Feature Request] nRF24L01 support

### ISSUE THREAD

**Issue by brandond** [esphome/feature-requests#45]
I have a use case for some small battery-powered distributed sensors. Ideally this would be a good use for BLE or [Z-Wave](http://www.z-wave.com/shop-z-wave-smart-home-products/category/smart-sensors), but BLE support has a ton of overhead and Z-Wave is [proprietary and expensive](https://www.amazon.com/dp/B01IRBDBZY).

What I'm probably going to end up doing is setting up a few devices with STM32s that will spend most of their time in suspend, occasionally waking to make a sensor reading and send it off via a [nRF24L01](https://platformio.org/lib/show/433/RF24) radio if conditions are met. On the ESP side, it would just be a matter of wiring up the RF24 SPI/Enable/IRQ pins to GPIOs. 

 The chip is limited in hardware to 32 byte message payloads. Perhaps something like this:
```c
struct payload {
  unit8_t type;
  char unit[3];
  char value[14];
  char name[14];
};
```

Ideally, esphomelib would not need to define a list of sensors ahead of time, but simply reformat and pass the messages through to MQTT.

**Comment by nekromant** [comment]
Hm, this would be useful. I happen to have my own pure-C optimized library for nrf24l01 (basically @ManiacBug 's arduino ported to pure C. I'll see if it compiles for esp8266 (it should). But it's pretty lowlevel.

**Comment by diogosalazar** [comment]
The mysensors/MySensors project handles communication between nrF24l01 modules and seems to be able to register devices dynamically.
ESP8266 is supported as a gateway.

Would it make sense to integrate with that project instead of implementing from scratch?

**Comment by nekromant** [comment]
As far as I got the nRF24L01 part, we have a huge bunch of hardware and stacks:
- Dumb sensors/boards using simple protocols. As simple as send/receive a predefined payload. (Should be easiest). 
- rf24mesh devices. rf24mesh doesn't specify any protocol, it's just read/write, no device method discovery/enumeration.
- mysensors that @diogosalazar mentioned. (That one is new to me, I'll have to check it out)
- My own libaura RPC over nRF24L01 ( https://gitlab.com/ncrmnt/aura-transport-rf24 ). 

Some nodes have encryption (e.g. car immobilizer tokens),  and that's a whole other layer of weirdness. I really want to help out with this task and have quite a lot of experience with nRF24L01 (non-arduino, pure C), but I don't know the inner workings of esphome good enough.

**Comment by glmnet** [comment]
I discovered ESPHome because I hated MySensors, their code quality seems ok, but these things communicating never was reliable for me, probably I was doing something wrong, I don't know.

As for today I believe this FR is outdated, I don't see need for dynamic sensor creation, though it is possible, in fact it is done already, you can just pick the message analyze it and push to any template sensor / binary sensor / cover... whatever, if that is what you need.

@brandond  What ended happening here?

**Comment by nekromant** [comment]
@glmnet I just learned about mysensors from this thread and rigged up a test mesh based on my older projects. (Yay, they're no longer sitting in the attic!). So far it looks like the best we have for nRF24L01. As for being outdated - I doubt that. nRF24L01 still provides better battery life for devices like leak sensors, compared to esp8266 and doesn't have a severely over-engineered std, like BLE. As for esphome, I think the only thing we can add - integrating a mysensors gateway in a few lines of yaml. For some cases this may be useful.

**Comment by glmnet** [comment]
It’s a great platform. It did not work for me. 

I just don’t know why would you use it within an esphome controller node. They provide a esp8266 gateway. That is you can connect hass to it then to my sensors network. It’s fully supported and documented.

**Comment by nekromant** [comment]
@glmnet  I guess you're right. The only thing I can think of - combining esphome node and mysensors gateway. As for non-mysensors protocols - perhaps nrf24l01 beacons and presense sensors come to my mind, but still with something like mysensors around, the whole idea of integrating nrf24l01 into esphome sounds like a waste of time (

**Comment by Niorix** [comment]
Any news? I have an esphome-based LED dimmer and I would have controlled it with a touch R11 remote controler (https://aliexpress.ru/wholesale?catid=&Searchtext=R11%20Remote). It's apparently uses 2.4. NRF. Therefore, I cannot directly connect him with Esphome((

**Comment by kluszczyn** [comment]
> The only thing I can think of - combining esphome node and mysensors gateway.

See: https://github.com/mannkind/ESPHomeMySensorsGatewayShim

**Comment by nekromant** [comment]
> > The only thing I can think of - combining esphome node and mysensors gateway.
> 
> See: https://github.com/mannkind/ESPHomeMySensorsGatewayShim

Thanks a ton! That's precisely what I've been looking for!

**Comment by otmezger** [comment]
I would love to be able to write yaml and leverage a nrf24l01 wireless network. let esphome do all the job ;-)

**Comment by jamesmyatt** [comment]
I wonder if this could be modelled on [`bluetooth_proxy`](https://esphome.io/components/bluetooth_proxy.html) and [BTHome ](https://bthome.io/).

**Comment by 3esmit** [comment]
I need nRF24L01 support for integrating a device that uses a remote control on 2.4GHz. I have captured the signals and I can send it using HackRF, but for integrating with home assistant would be nice to have this support.

**Comment by jhogendorn** [comment]
> I need nRF24L01 support for integrating a device that uses a remote control on 2.4GHz. I have captured the signals and I can send it using HackRF, but for integrating with home assistant would be nice to have this support.

I am also keen to get this use case working

**Comment by zackfuchtel** [comment]
[x] AOL

**Comment by sudipta26889** [comment]
ESPHome core doesn't have nRF24L01 support yet. right?

Do we have any external component?

**Comment by derGreyson** [comment]
Is there any option to fix this manual?

**Comment by AdamM68** [comment]
I'm still waiting for nRF24L01 support in esphome and home assistant.
Do Yoy know any working solution ?

**Comment by Lukanex** [comment]
Hi, I would also like to use nRF24L01 or RMF73 with Home Assisntant (preferrably via ESPHome).
Any reccomendations, please?

I have looked at the [MySensors ](https://www.mysensors.org/) platform and it coveres some of my use-cases and has integration with HA. The need to manage another code base is one of the down-sides for me.

**Comment by jamesmyatt** [comment]
What do you imagine nRF24L01 support in ESPHome to look like? nRF24L01 is just the radio, i.e. the physical layer.

**Comment by Lukanex** [comment]
> What do you imagine nRF24L01 support in ESPHome to look like? nRF24L01 is just the radio, i.e. the physical layer.

That is a very good question.
For my first purpose: to be able to setup a custom network with one gateway and multiple sensors. Star, mashing not required.

But, now that I am thinking about it, it would be best to adhere to some standard protocol to really bring a value to the ESPHome users, right? I looked up that Enhanced ShockBurs (ESB) is common amongst hobbyists.

I am on the fence with MySensors. They support [ESP8266 and ESP32](https://github.com/mysensors/MySensors/tree/development/hal/architecture), none of the newer chips.
It would be super convenient to manage the MySensors nodes via the ESPHome interface :)

**Comment by jhogendorn** [comment]
> What do you imagine nRF24L01 support in ESPHome to look like? nRF24L01 is just the radio, i.e. the physical layer.

My interest is in replacing RF based remotes, although this is more complex than just rf cloning.

**Comment by nekromant** [comment]
Since github's pinging me about the issue, let me throw my 2 cents in: 
 
@jhogendorn These seldom use nrf24l01 - too expensive for that purpose, mostly some dumb 433 or similar transiever ICs. These are often used in keyboards/mouses - but getting those supported in esphome is (apart from being a fun pasttime) doesn't really have any use. 

In terms of mysensors - as @kluszczyn noticed, https://github.com/mannkind/ESPHomeMySensorsGatewayShim gets most of the job done. Perhaps some management via web ui might be useful, but that (IMO) would have to be done outside esphome. 

Another idea I had flying around - implement a esphome-like tool for building mysensors nodes(Man,. I wish I had the time!). But that's totally out of scope of this project.

**Comment by jhogendorn** [comment]
> [@jhogendorn](https://github.com/jhogendorn) These seldom use nrf24l01 

Nah, i have a bunch of remotes that are utilising 2.4ghz chipsets, i've tried doing some sniffing on them with a hackrf but i'm somewhat out of my depth.

**Comment by nekromant** [comment]
@jhogendorn For that case nrf24l01 component will make sense. 

Check out this hack https://github.com/Yveaux/NRF24_Sniffer This may be a good start for you

**Comment by Entropy512** [comment]
> These seldom use nrf24l01

Nope.  Logitech Harmony, one of the most commonly known RF remotes on the market (well, formerly on the market...) uses Nordic's Enhanced Shockburst protocol (supported by the NRF24L01 and Nordic's integrated RF SoCs).

https://github.com/joakimjalden/Harmoino - At least my envisioned use case would be an ESPhome-integrated version of this.

**Comment by jamesmyatt** [comment]
I assume that the packet transport system will do for this, like ESP-NOW: https://esphome.io/components/packet_transport/

BTW, I've seen wireless mice and keyboards using nrf24l01.

**Comment by loucks1** [comment]
I have copied the nrf24/RF24 driver into an Esphome external component - I had to implement this to get my directolor solution to compile using esp-idf instead of the arduino I had been using.  I can read and send, so I think it is starting to get to a point where I can share it.  If someone has time to help me with it, I think we could whip it into shape pretty quickly.  I tried to keep it as close to nrf24/RF24 as possible.  The key point here is that I'm using the esphome spi and spi_device.  I would like to run a test with it use nrf24/RF24 to make sure that I can write and send to something using the official library.  If you want to help, reach out to me directly.  I think it is a long way from being ready to be included officially in esphome.

**Comment by loucks1** [comment]
I haven't implemented the samples yet, but here is the repository that works for my directolor component:  https://github.com/loucks1/nrf24_espidf

**Comment by charly4711** [comment]
> I haven't implemented the samples yet, but here is the repository that works for my directolor component: https://github.com/loucks1/nrf24_espidf

I'll be looking at that in order to use a DX-DTU device to read Hoymiles microconverter data via ESP32->SPI->NRF24L01+ 
I would really prefer using that with esphome and hope a bit of duct-tape, this, and the Hoymiles library in OpenDTU could do the trick. But it may take me a while, given that I have C and Java xp, but not C++.

**Comment by loucks1** [comment]
> I'll be looking at that in order to use a DX-DTU device to read Hoymiles microconverter data via ESP32->SPI->NRF24L01+ I would really prefer using that with esphome and hope a bit of duct-tape, this, and the Hoymiles library in OpenDTU could do the trick. But it may take me a while, given that I have C and Java xp, but not C++.

Why not just use the OpenDTU and MQTT? 

MQTT support (with TLS)
Home Assistant MQTT Auto Discovery support

They've already done the hard work.  Unless you want to learn the nrf24 protocol, I highly recommend you just go with OpenDTU - you'll have a much better experience and you can be up in a couple of hrs with AI helping you.  If you go this route, you're talking weeks probably.

**Comment by charly4711** [comment]
> Why not just use the OpenDTU and MQTT?
> 
> MQTT support (with TLS) Home Assistant MQTT Auto Discovery support
> 
> They've already done the hard work. Unless you want to learn the nrf24 protocol, I highly recommend you just go with OpenDTU - you'll have a much better experience and you can be up in a couple of hrs with AI helping you. If you go this route, you're talking weeks probably.

It's a combination of me being paranoid and wanting to minimize the attack vectors by eliminating web UI and API, making sure I get WPA3-only Wifi, etc. I'm also not in a hurry and it is a bit of a learning project. But you're right, me giving up at some point and just taking the working OpenDTU solution is totally possible.
